OpenRA.Mods.Cnc/FileSystem/MegFile.cs

```diff
@@ -0,0 +1,206 @@
+		{
+			private readonly Stream s;
+
+			private readonly List<string> contents = new List<string>();
+
+			private readonly List<MegFileContentReference> fileData = new List<MegFileContentReference>();
+			private readonly string name;
+
+			internal MegFile(FileStream s)
+			{
+				this.s = s;
+
+				var reader = new BinaryReader(s);
+
+				var id1 = reader.ReadUInt32();
+				var id2 = reader.ReadUInt32();
+
+				if (!id1.Equals(HeaderId1) || !id2.Equals(HeaderId2))
+					throw new Exception("Invalid file signature for meg file");
+
+				ParseMegHeader(reader);
+			}
+
+			public MegFile(Stream s, string filename)
+			{
+				this.s = s;
+				name = filename;
+
+				ParseMegHeader(new BinaryReader(s));
+			}
+
+			/// <summary>
+			/// This method reads the file tables from a file. It is assumed the header magic bytes have already been read.
+			/// </summary>
+			/// <param name="reader">The reader of the stream</param>
+			private void ParseMegHeader(BinaryReader reader)
+			{
+				var dataStartOffset = reader.ReadUInt32();
+				var numFileNames = reader.ReadUInt32();
+				var numFiles = reader.ReadUInt32();
+				var fileNameTableSize = reader.ReadUInt32();
+
+				// the file names are an indexed array of strings
+				for (uint i = 0; i < numFileNames; i++)
+				{
+					var fileNameLength = reader.ReadUInt16();
+					var fileNameBytes = reader.ReadBytes(fileNameLength);
+					var fileName = Encoding.ASCII.GetString(fileNameBytes);
+
+					contents.Add(fileName);
+				}
+
+				// the header indicates where we should be, so verify it
+				if (reader.BaseStream.Position != fileNameTableSize + HeaderFileNameTableStartOffset)
+					throw new Exception("File name table in .meg file inconsistent");
+
+				// now we load each file entry and associated info
+				for (var i = 0; i < numFiles; i++)
+				{
+					var flags = reader.ReadUInt16();
+					if (flags != 0) throw new Exception("Encrypted files are not supported or expected.");
+
+					var crc = reader.ReadUInt32();
+					var fileRecordIndex = reader.ReadUInt32();
+					var fileSize = reader.ReadUInt32();
+					var fileStartOffset = reader.ReadUInt32();
+					var fileNameIndex = reader.ReadUInt16();
+
+					fileData.Add(new MegFileContentReference(crc, fileRecordIndex, fileSize, fileStartOffset, fileNameIndex));
+				}
+
+				if (reader.BaseStream.Position != dataStartOffset)
+				{
+					throw new Exception("Expected to be at data start offset");
+				}
+			}
+
+			public string Name { get { return name; } }
+
+			public IEnumerable<string> Contents { get { return contents; } }
+
+			public IReadOnlyDictionary<string, PackageEntry> Index
+			{
+				get
+				{
+					var absoluteIndex = Contents.ToDictionary(e => e, e =>
+					{
+						var index = contents.IndexOf(e);
+						var reference = fileData.FirstOrDefault(a => a.FileNameTableRecordIndex == index);
+						return new PackageEntry(reference.Crc32, reference.FileStartOffsetInBytes, reference.FileSizeInBytes);
+					});
```

Comment:
Style nit: empty line after blocks (statements formatted like a block, as in this case).